# Tutorial: Core Metrics with `sklearn.metrics`

Setelah kita melatih model dan mendapatkan prediksi, kita perlu cara objektif untuk menjawab pertanyaan: **"Seberapa besar kesalahan prediksi model kita?"**.

Di notebook ini, kita akan fokus pada tiga metrik evaluasi fundamental untuk regresi. Semuanya tersedia di modul `sklearn.metrics`. Kita akan belajar cara mengimplementasikan dan, yang lebih penting, cara menginterpretasikan masing-masing metrik.

**Metrik yang akan kita pelajari:**
1.  **Mean Absolute Error (MAE)**
2.  **Mean Squared Error (MSE)**
3.  **Root Mean Squared Error (RMSE)**

### 1. Setup: Melatih Model untuk Dievaluasi

Sebelum kita bisa mengevaluasi, kita perlu model dan prediksi. Mari kita ulangi alur kerja standar dengan cepat menggunakan dataset "Advertising".


In [15]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression

# memuat Data
data = {
    'TV': [230.1, 44.5, 17.2, 151.5, 180.8, 8.7, 57.5, 120.2, 8.6, 199.8],
    'Radio': [37.8, 39.3, 45.9, 41.3, 10.8, 48.9, 32.8, 19.6, 2.1, 2.6],
    'Newspaper': [69.2, 45.1, 69.3, 58.5, 58.4, 75.0, 23.5, 11.6, 1.0, 21.2],
    'Sales': [22.1, 10.4, 12.0, 16.6, 17.9, 7.2, 11.8, 13.2, 4.8, 15.6]
}
df = pd.DataFrame(data)

# 1. pisahkan data fitur (X) dan target (Y)
X = df.drop(columns=['Sales'])
Y = df['Sales']

# 2. lakukan tran test split
X_train, X_test, Y_train, Y_test = train_test_split(X,Y,test_size=0.2,random_state=42)

# 3. latih model 
model = LinearRegression()
model.fit(X,Y)

# 4. prediksi data test
Y_pred = model.predict(X_test)

---
### 2. Mean Absolute Error (MAE)

**Intuisi:** MAE menjawab pertanyaan, **"Rata-rata, seberapa jauh prediksi kita dari nilai sebenarnya?"**. Ini adalah metrik yang paling mudah dipahami secara langsung.

**Cara Kerja:** Ia menghitung selisih absolut (tanpa nilai negatif) antara setiap prediksi dan nilai sebenarnya, lalu mengambil rata-ratanya.

`MAE = (1/n) * Σ |y_aktual - y_prediksi|`

In [16]:
from sklearn.metrics import mean_absolute_error

# hitung MAE
mae = mean_absolute_error(Y_test,Y_pred)

print(f'Mean Absolute Error (MAE) : {mae:.2f}')

Mean Absolute Error (MAE) : 0.24


#### Interpretasi MAE:

Nilai MAE **0.24** berarti bahwa, secara rata-rata, prediksi penjualan kita **salah sekitar 0.24 unit**. Jika penjualan diukur dalam ribuan, maka kesalahan rata-rata kita adalah 0.24 unit. Metrik ini sangat bagus untuk dilaporkan ke pemangku kepentingan bisnis karena unitnya mudah dipahami.


### 3. Mean Squared Error (MSE)
**Intuisi:** MSE juga mengukur kesalahan rata-rata, tetapi dengan satu perbedaan penting: ia **mengkuadratkan selisihnya** sebelum dirata-ratakan.

**Cara Kerja:** Proses pengkuadratan ini memberikan **hukuman yang jauh lebih besar untuk kesalahan yang besar**. Sebuah prediksi yang salah 10 unit akan berkontribusi 100 pada total kesalahan, sedangkan prediksi yang salah 2 unit hanya berkontribusi 4.

`MSE = (1/n) * Σ (y_aktual - y_prediksi)²`


#### Interpretasi MSE:
Nilai MSE **3.6792** sulit diinterpretasikan secara langsung karena unitnya adalah **"unit penjualan kuadrat"**, yang tidak intuitif. Kegunaan utama MSE adalah sebagai "fungsi kerugian" (*loss function*) yang sering dioptimalkan oleh algoritma ML karena sifat matematisnya yang baik. MSE sangat berguna jika Anda ingin memastikan model Anda tidak membuat kesalahan yang sangat besar.

### 4. Root Mean Squared Error (RMSE)

**Intuisi:** RMSE adalah **akar kuadrat dari MSE**. Tujuannya sederhana: mengembalikan unit dari MSE kembali ke unit asli target, sehingga lebih mudah diinterpretasikan.

**Cara Kerja:** Ia memiliki properti yang sama dengan MSE (menghukum kesalahan besar), tetapi hasilnya sebanding dengan MAE. RMSE dapat dianggap sebagai **"standar deviasi dari kesalahan prediksi"**.


#### Interpretasi RMSE:

Nilai RMSE **1.9181** berarti bahwa **kesalahan tipikal** atau standar deviasi dari kesalahan prediksi model kita adalah sekitar **1.91 unit penjualan**.
**Perbandingan MAE vs. RMSE:**
* Perhatikan bahwa `RMSE (1.91)` sedikit lebih besar dari `MAE (1.51)`.
* Ini hampir selalu terjadi karena RMSE memberikan bobot lebih pada kesalahan yang lebih besar.
* Perbedaan yang besar antara RMSE dan MAE menandakan bahwa model Anda membuat beberapa kesalahan yang sangat besar (*outlier errors*).

### 5. Ringkasan dan Kapan Menggunakannya

| Metrik | Kelebihan | Kekurangan | Kapan Digunakan? |
| :--- | :--- | :--- | :--- |
| **MAE** | Paling mudah diinterpretasikan, tidak sensitif terhadap outlier. | Tidak menghukum kesalahan besar. | Untuk melaporkan performa model ke audiens non-teknis. |
| **MSE** | Menghukum kesalahan besar, bagus untuk optimasi. | Sulit diinterpretasikan, unitnya kuadrat. | Sebagai *loss function* internal selama pelatihan model. |
| **RMSE** | Menghukum kesalahan besar, unitnya dapat diinterpretasikan. | Lebih sensitif terhadap outlier daripada MAE. | Metrik standar di banyak kompetisi dan laporan teknis. |

**Selanjutnya:**
Metrik-metrik ini memberi tahu kita besarnya kesalahan dalam **satuan absolut**. Di subchapter berikutnya, kita akan mempelajari **R-squared (R²)**, yang memberi tahu kita seberapa baik model kita dalam **persentase relatif**.
